src/DurableTask.Netherite/StorageLayer/Faster/AzureBlobs/AzureStorageDevice.cs

```diff
@@ -516,31 +556,83 @@ async Task ReadFromBlobAsync(UnmanagedMemoryStream stream, BlobUtilsV12.PageBlob
+
+                                    using (var streamingResult = response.Value)
+                                    {
+                                        await streamingResult.Content.CopyToAsync(stream).ConfigureAwait(false);
+                                    }
+
+                                    // we have observed that we may get 206 responses where the actual length is less than the requested length
+                                    // this seems to only happen if blob was enlarged in the past
+                                    long actualLength = (stream.Position - offset); 
+
+                                    if (actualLength < requestedLength)
+                                    {
+                                        this.BlobManager.StorageTracer?.FasterStorageProgress($"PageBlob.DownloadStreamingAsync id={id} returned partial response range={response.Value.Details.ContentRange} requestedLength={requestedLength} actualLength={actualLength}");
+                                       
+                                        if (actualLength == 0)
+                                        {
+                                            throw new InvalidDataException($"PageBlob.DownloadStreamingAsync returned empty response, range={response.Value.Details.ContentRange} requestedLength={requestedLength} ");
+                                        }
+                                        else if (actualLength % 512 != 0)
+                                        {
+                                            throw new InvalidDataException($"PageBlob.DownloadStreamingAsync returned unaligned response, range={response.Value.Details.ContentRange} requestedLength={requestedLength} actualLength=${actualLength}");
+                                        }
+                                        else
+                                        {
+                                            length = (uint)actualLength; // adjust length to actual length read so the next read will start where this read ended
+                                        }
+                                    }
+                                    else if (actualLength > requestedLength)
+                                    {
+                                        throw new InvalidDataException($"PageBlob.DownloadStreamingAsync returned too much data, range={response.Value.Details.ContentRange} requestedLength={requestedLength} actualLength=${actualLength}");
+                                    }
+                                }
+                                catch (Azure.RequestFailedException e) when (e.ErrorCode == "InvalidRange")
+                                {
+                                    // this type of error can be caused by the page blob being smaller than where FASTER is reading from.
+                                    // to handle this, first determine current page blob size.
```

Comment:
not sure I fully understand this comment. What does it mean to be smaller than _where_ FASTER is reading from? Can we add more detail?

Reply:
FASTER may be reading a range that extends past the end of the blob.